### PR TITLE
feature(manual cancellation update): Reset loading state when outstanding call is cancelled

### LINKED ISSUE #372: 'loading' stays true when request with 'manual' option cancels
In the below code example, if I call 'refetch' manually and it cancels (without a new 'refetch' happening) - then the loading state variable will stay true.

```
  const [{ data, loading }, refetch] = useAxios(
    {
      url: `/someUrl?param=${iCanChange}`
    },
    {
      manual: true
    }
  )
```

I think what causes the bug:
The refetch cancels because the 'iCanChange' param changes  (correct behavior).
No new 'refetch' is initiated because of the 'manual' option (correct behavior).
But now the 'loading' state stays true (false behavior?)

### PR DESCRIPTION
**Pull request by vmlopezr**
Here's a proposal for #372 

1. Resets the loading property (and keeps previous state) when a outstanding call is cancelled via the manualCancel method. 
   - This required tracking how the `cancelOutstandingRequest` is called with the `isCancelledManually` ref. The state update only occurs when the `cancelManually` method is executed.

2. Resets the loading property when a manual call is cancelled via a config change to the hook (on new line 375). 
   - Required checking that the component is mounted as it attempts to update the state on a config change (that is also triggered on an unmount). This was done with the `isMounted` ref and the empty dependency useEffect.
 
    *Note:* Point 2 is done under the assumption that changes in the config and option should cancel a outstanding request when the manual property is true per #372.

The sandbox shows the operation of the PR. Clicking the `cancel request` button will reset the `loading` property and keep the previous state. Likewise, updating the hook config will cancel and reset the `loading` property.

https://codesandbox.io/s/axios-hooks-manual-cancel-updates-wvv4z?file=/src/App.js

Closes #372

### CODE REVIEW

**Comment by vmlopezr** [comment]
Looking to hear what you think. The sandbox is what i wanted to use for the example on the manual cancel PR (minus the unmount button) before i noticed that the loading property would stay `true`.

With that in mind, i found it tough to figure out how to test the cancelling: 
1. Trigger request that takes some time (say 1s), cancel it before timeout and expect the results to be the previous state and for loading to be false (codecov on line 156)
2. Test that the hook state does not update when the component unmounts (codecov 197-198).

**Review by kentongray — APPROVED** [review]

**Comment by vmlopezr** [comment]
Hey @simoneb  just wanted to check if you had any comments or suggestions to improve this.

**Comment by simoneb** [comment]
@vmlopezr thanks for this and sorry for the delay. I see quite a bit of additional complexity which I'm not sure is necessary, I'm wondering if we could simplify this implementation by always resetting the loading property when a cancel happens. Also, because we're basically introducing cancellation as a first class citizen, I suspect we could model that more explicitly as far as state handling (i.e. the reducer) is concerned, e.g. with an explicit action, instead of reusing the REQUEST_END action

**Comment by simoneb** [comment]
I think we need a state diagram otherwise reasoning about this is extremely difficult. 

> So, i think that means that we should only be resetting the loading property on a manual cancel, and not on refetches. Any ideas? (This kind of brings it back to the ref that i passed on my initial commits)

This kinda makes sense, yes. In any case we should always think about how we **want** this to behave, before thinking about the implementation.

When a cancellation occurs because of refetch, I believe you don't want to see the loading flag flickering because you know another fetch is going to happen. 

When canceling due to manual cancellation, I guess it makes sense to set loading to false instead, because we don't know whether a refetch will happen, and if it will it will set loading to true anyway.

Makes sense? Anyway, a state chart would be great to have.

**Comment by vmlopezr** [comment]
@simoneb Not sure if this is what you wanted, but I made a rough diagram of the state changes. 
As I see it, the loading state would be reset when:
 - A request finishes successfully or with an error (not counting cancellation from refetches)
 - Cancellation:
   - Called manually through manual cancel call.
   - A config change when manual is true. (This was the issue from #372 )

axios-hooks-state

**Review by simoneb — APPROVED** [review]
I'm going to merge this even though it introduces a change in the existing behavior. When a cancel happens because of configuration change on a non-manual request, the loading flag used to stay true whereas it will go on false with this change to then return to true when the next request is triggered.

Nevertheless, I like the simplicity of this implementation and the conceptual correctness of the new behavior. Thanks for this contribution, very valuable.

**Comment by vmlopezr** [comment]
@simoneb Just saw that this got merged yesterday.

Per your comment `the loading flag used to stay true whereas it will go on false with this change to then return to true when the next request is triggered.` 

Just wanted to verify that you understood that when triggering multiple calls (so that the hook cancels the outstanding request), the loading property resets to false until the most recent outstanding request resolves. It **doesn't return to true** on refetch.

See https://codesandbox.io/s/axios-hooks-manual-cancel-updates-forked-fif0i?file=/src/App.js  and click fast on either the 'external refetch' or 'state refetch' buttons three times in a row.

You'll see `...loading` show up on the first click and then go away on the second click.  The reason for this is that the `catch` executes only after the new call is triggered. That is, the `loading` property ends up toggling from `true`(on the refetch) to false (on the `catch`)

The diagram visualized what should happen (basically a diagram of the first commits of this PR) and was actually not a visualization of the recent commit that resets the loading property on cancel.

Since people expect the 'loading' property to show whether the call is in progress, i imagine there might be an issue about this in the future.

**Comment by simoneb** [comment]
I'm going to revert this change then.

**Comment by vmlopezr** [comment]
@simoneb 

Regarding issue #372,  the initial commits before your first reply captured the intended execution. 
- When a request is cancelled manually, reset the loading property
- Otherwise, don't update state on cancellation.

The only way i could think of figuring out the cancellation happened manually was to keep a ref, and pass it to the refetch method. 

Here's an sandbox of the intended functionality .
https://codesandbox.io/s/axios-hooks-manual-cancel-updates-forked-0lsqq?file=/src/axios-hooks.js

**Comment by simoneb** [comment]
Thanks @vmlopezr and apologies for the rush in getting this out, I missed that the implementation was not complete. I'd like to have a better look at it because I have a feeling that there may be a better solution at the code level, but I'm not 100% sure and I also don't have too much time to dedicate to this. 

Basically, I would like to use an approach that relies on the cancellation reason being provided when a request is cancelled, and then use that one to decide what to do with the loading flag, instead of having to rely on additional hooks as per your solution. I believe I understand your point about the cancellation happening too late, which would make the approach I'm proposing unfeasible, but I would nevertheless like to see if this would be doable, even at the cost of changing how the hook is implemented right now.

The reason is that there is quite a bit of logic which is already hard to follow in the current implementation, and I would prefer to avoid increasing the complexity further.

**Comment by vmlopezr** [comment]
@simoneb This comment is just brainstorming.

But i just thought of this; we could still have the reducer action `actions.REQUEST_CANCELED`. 

The request catch would stay as is
```js
    try {
       ....
    } catch (err) {
      if (!StaticAxios.isCancel(err)) {
        dispatch({ type: actions.REQUEST_END, payload: err, error: true });
      } 
      // If cancelled don't do anything else
      throw err;
    }
```

Instead, on the `manualCancel` return we can do this
```jsx
    const cancelManually = React.useCallback(() => {
     // Cancel the outstanding request
      cancelOutstandingRequest();
     // Reset loading state
      dispatch({ type: "REQUEST_CANCELLED" });
    }, [cancelOutstandingRequest]);
```
This makes it so that the state is reset when the user cancels manually, and removes the additional logic(the refs) i showed you initially, , i.e. there's no need to do anything else on the catch of the cancelled request.

See the sandbox below.
https://codesandbox.io/s/axios-hooks-manual-cancel-updates-forked-0lsqq?file=/src/axios-hooks.js

Let me know what you think of that. If not i'll try to think about it from another direction.

**Comment by GitKokoszka** [comment]
@simoneb Is there any update about this issue? I just encountered it and it's nasty in production code - I need to create separate state for loading and create new function that fires current cancelRequest implementation and resets my additional loading flag.

**Comment by simoneb** [comment]
@GitKokoszka I reopened https://github.com/simoneb/axios-hooks/issues/372 but there is no solution yet.
